lib/apiservers/engine/backends/archive.go

```diff
@@ -487,13 +490,92 @@ func (rm *ArchiveStreamReaderMap) FindArchiveReaders(containerSourcePath string)
 		}
 
 		if !found {
-			nodes = append(nodes, startingNode)
+			// prepend the starting node at the beginning
+			nodes = append([]*ArchiveReader{startingNode}, nodes...)
 		}
+	} else {
+		startingNode = nodes[0]
+	}
+
+	err = rm.buildFilterSpec(containerSourcePath, nodes, startingNode)
+	if err != nil {
+		return nil, err
 	}
 
 	return nodes, nil
 }
 
+func (rm *ArchiveStreamReaderMap) buildFilterSpec(containerSourcePath string, nodes []*ArchiveReader, startingNode *ArchiveReader) error {
+	var err error
+
+	// Build an exclusion filter for each writer.  For example, the reader for / should not read
+	// from submounts as there are separate readers for those.
+	buildExclusion := func(path string, node *ArchiveReader) error {
+		var walkResults []string
+
+		childWalker := func(prefix patricia.Prefix, item patricia.Item) error {
+			if _, ok := item.(*ArchiveReader); !ok {
+				return fmt.Errorf("item not ArchiveReader")
+			}
+
+			ar, _ := item.(*ArchiveReader)
+			walkResults = append(walkResults, ar.mountPoint.Destination)
```

Comment:
Why not the following instead of walkResults?
```
dest := ar.MountPoint.Destination
if path != dest {
  node.filterSpec.Exclusions[dest] = struct{}{}
}
```

We should also make sure that `dest` has a `/` suffix - that way the mountpoint itself can be `stat`'d, for example, but files under it will not be included.